mmengine/hooks/profiler_hook.py

```diff
@@ -0,0 +1,216 @@
+import os.path as osp
+import warnings
+from typing import Callable, Optional, Union
+
+import torch
+
+from mmengine.dist import master_only
+from mmengine.hooks import Hook
+from mmengine.registry import HOOKS
+
+
+def check_kineto() -> bool:
+    kineto_exist = False
+    try:
+        if torch.autograd.kineto_available():
+            kineto_exist = True
+    except AttributeError:
+        warnings.warn('NO KINETO')
+    return kineto_exist
+
+
+@HOOKS.register_module()
+class ProfilerHook(Hook):
+    """ProfilerHook to analyze performance during training.
+
+    PyTorch Profiler is a tool that allows the collection of the performance
+    metrics during the training. More details on Profiler can be found at
+    https://pytorch.org/docs/1.13.1/profiler.html#torch.profiler.profile
+
+    Args:
+        by_epoch (bool): Profile performance by epoch or by iteration.
+            Default: True.
+        profile_times (int): The period (epoch/iter) recorded by the profiler.
+            Eg: profile_iters=10 and by_epoch=False, record 0-10 iteration.
+            Default: 1.
+        activity_with_cpu (bool): Activities to be used in the analysis (CPU)
+        activity_with_cuda (bool): Activities to be used in the analysis (CUDA)
+        schedule (dict, optional): Config of generating the callable schedule.
+            The dict can include wait、warmup、active、repeat、skip_first.
+            Default: None, mean not add step markers
+        on_trace_ready (callable, dict, optional): Either a handler or a dict
+            of generate handler.
+            [Terminal] dict(type='log_trace')
+            [Tensorboard] dict(type='tb_trace', **trace_cfg)
+                trace_cfg include dir_name、worker_name、use_gzip
+                dir_name default to "{work_dir}/tf_tracing_logs".
+            Default: None, mean
+        record_shapes (bool): Save information about operator's input shapes.
+            Default: False.
+        profile_memory (bool): Track tensor memory allocation/deallocation.
+            Default: False.
+        with_stack (bool): Record source information (file and line number)
+            for the ops.
+            Default: False.
+        with_flops (bool): Use formula to estimate the FLOPS of specific
+            operators (matrix multiplication and 2D convolution).
+            Default: False.
+        json_trace_path (str, optional): Exports the collected trace in Chrome
+            JSON format. Chrome use 'chrome://tracing' view json file.
+            Default: None, mean don't save json.
+    Examples:
+        >>> # tensorboard trace
+        >>> trace_config = dict(type='tb_trace', dir_name='work_dir')
+        >>> profiler_hook_cfg = dict(on_trace_ready=trace_config)
+    """
+
+    def __init__(self,
+                 by_epoch: bool = True,
+                 profile_times: int = 1,
+                 activity_with_cpu: bool = True,
+                 activity_with_cuda: bool = False,
```

Comment:
Suggest using `activities` that support strings and torch.profiler.ProfilerActivity. 'cpu' -> torch.profiler.ProfilerActivity.CPU; 'cuda' -> torch.profiler.ProfilerActivity.CUDA.

Reply:
torch.profiler is collected by means of the set two fixed value (torch. Profilers. ProfilerActivity. CPU, torch. The profiler. ProfilerActivity. CUDA). I feel like it would be unintuitive to use a list of Strings.

Reply:
![1670807496399](https://user-images.githubusercontent.com/34785944/206941168-a8a882dd-4647-43eb-aded-67dd6b0e845f.png)